Display a status page for translations

viroulep opened the pull request:
Fixes #847.

http://internationalize.herokuapp.com/ will add a hash of the original value to the translated keys in the file, ie it will output something like this:

```yml
fr:
  countries:
    #original_hash: 3f8c9de
    XA: Pays multiple (Asie)
    #original_hash: 090f3ce
    XE: Pays multiple (Europe)
```

To parse contextual comment in the original file, Jonatan uses some regexp on the whole file, at every key of the file (see the original code [here](https://github.com/viroulep/internationalize/blob/b33561347e0f67fb8b0f4d57f6a42034b93dbc95/client/app/services/translation-utils.service.js#L194-L220)).

It works pretty fast there, I adapted it in ruby (see [here](https://github.com/thewca/worldcubeassociation.org/pull/1090/files#diff-42ebd3a672fdd4205143d95c39ddd848R27)) and it worked ok for the small locale files I tested it on.
However I just tried using it on the current French one, and it's insanely slow (ie: not usable!).
I'm not aware of the performance differences between regexp in js and in ruby, if any.
Some questions for you:
 - is there a way to improve the regexp speed? FYI right now the French locale generated ~500 regexp on a ~30kB text.
 - I'm very very tempted to build a very small parser that would iterate through the file line per line, making it ~500 very simple regexp on ~60B text, do you think it would be a good alternative?
 - Is extending the ruby YAML library a good idea?

Feedback on other part of the code is welcome too, even if it's still a wip :)

Here is a screenshot to have an idea of what it looks like:
![status-locale](https://cloud.githubusercontent.com/assets/1007485/21870544/2b529b3c-d85d-11e6-9b37-ada23805e4e7.png)

jonatanklosko commented on WcaOnRails/app/controllers/server_status_controller.rb:
```diff
@@ -8,6 +8,14 @@ def index
+    (I18n.available_locales - [:en]).each do |l|
+      ref_locale = Locale.new(l, true)
+      missing, outdated = ref_locale.compare_to(@ref_english)
+      @status_locales[l] = { "missing": missing, "outdated": outdated }
```
What about using symbols?

jonatanklosko commented on WcaOnRails/app/models/locale.rb:
```diff
@@ -0,0 +1,124 @@
+
+  def initialize(locale, is_translation=false)
+    self.locale = locale.to_s
+    filename = Rails.root.join('config', 'locales_test', "#{locale}.yml")
```
I'd suggest moving those files to the specs directory, e.g. support.

viroulep replied:
This is actually only for debug purpose: I intend to remove the `locales_test` directory and directly load the real locale :)
I just wanted to mimic the real behavior with "fake" locales.

jonatanklosko commented on WcaOnRails/app/models/locale.rb:
```diff
@@ -0,0 +1,124 @@
+        # $2 is the hash(es) matched, or empty
+        hashes = $2
+        unless hashes.empty?
+          hash_lines = hashes.split('#').map { |l| l.strip!() }
```
`.map!(&:strip)`

jonatanklosko commented on WcaOnRails/app/models/locale.rb:
```diff
@@ -0,0 +1,124 @@
+          hash_lines = hashes.split('#').map { |l| l.strip!() }
+          if hash_lines
+            hash = hash_lines.select(&method(:line_filter)).map(&method(:line_cleaner))
+            if hash.size() >= 1
```
These parentheses are not necessary, you've it in many places, so I mentione it just here.

jonatanklosko commented on WcaOnRails/app/models/locale.rb:
```diff
@@ -0,0 +1,124 @@
+    missing_keys = []
+    outdated_keys = []
+    base.each do |key, value|
+      fully_qualified = (context + [key]).join("> ")
```
Don't you think that `" > "` would look better?

viroulep commented:
Thanks for the comments, I'll fix them and make rubocop happy later :)

I realized I could simply run the ruby file on its own and to some profiling, here is result of one single build of the first 179 lines of the French locales (there is a bunch more lines, but I don't think they're relevant for this issue):

>  %   cumulative   self              self     total
 time   seconds   seconds    calls  ms/call  ms/call  name
 91.58     1.74      1.74      163    10.67    10.67  String#gsub
  1.05     1.76      0.02       93     0.22     2.47  Array#each
  1.05     1.78      0.02      114     0.18     3.42  Kernel#require
  0.53     1.79      0.01       80     0.13     0.13  IO#set_encoding
  0.53     1.80      0.01      491     0.02     0.02  Module#method_added
  0.53     1.81      0.01      281     0.04     0.04  Psych::TreeBuilder#scalar
  0.53     1.82      0.01        1    10.00    10.00  Psych::Visitors::ToRuby#initialize
  0.53     1.83      0.01      281     0.04     0.04  Psych::Visitors::ToRuby#resolve_class
  0.53     1.84      0.01      281     0.04     0.11  Psych::Visitors::ToRuby#visit_Psych_Nodes_Scalar
  0.53     1.85      0.01      306     0.03     1.01  Psych::Visitors::ToRuby#accept
  0.53     1.86      0.01      152     0.07     0.07  Kernel#is_a?
  0.53     1.87      0.01      152     0.07     0.07  Regexp#initialize
  0.53     1.88      0.01      349     0.03    27.16  Locale#decorate_with_hashes
  0.53     1.89      0.01      281     0.04     0.07  Psych::Visitors::ToRuby#deserialize
  0.53     1.90      0.01      306     0.03     0.98  Psych::Visitors::Visitor#visit

So it looks like gsub is indeed the bottleneck here :s
And it's the gsub [here](https://github.com/thewca/worldcubeassociation.org/pull/1090/files#diff-42ebd3a672fdd4205143d95c39ddd848R46), not [there](https://github.com/thewca/worldcubeassociation.org/pull/1090/files#diff-42ebd3a672fdd4205143d95c39ddd848R118), because I removed the latter to be sure.

viroulep commented:
The `gsub`!!!
I just read the doc and gsub go through **all** the matches, when the regexp is actually carefully built to only need the first!

Here is the same profiling with `sub`, on the *whole* French locale this time:
>   %   cumulative   self              self     total
 time   seconds   seconds    calls  ms/call  ms/call  name
 14.74     0.14      0.14      695     0.20     0.22  String#sub
  8.42     0.22      0.08     1546     0.05     1.51  Locale#decorate_with_hashes
  8.42     0.30      0.08      433     0.18     3.70  Kernel#require
  6.32     0.36      0.06     5433     0.01     0.04  Gem::StubSpecification#data
  5.26     0.41      0.05     4351     0.01     0.01  Gem::StubSpecification#name
  5.26     0.46      0.05     2029     0.02     0.12  Gem::Specification._resort!
  4.21     0.50      0.04     1392     0.03     0.80  Psych::Visitors::Visitor#visit
  3.16     0.53      0.03      298     0.10     0.10  Gem::StubSpecification::StubLine#initialize
  3.16     0.56      0.03     1392     0.02     0.83  Psych::Visitors::ToRuby#accept
  3.16     0.59      0.03      695     0.04     0.04  Regexp#initialize
  2.11     0.61      0.02     2947     0.01     0.21  Class#new
  2.11     0.63      0.02     1392     0.01     0.01  Delegator.const_missing
  2.11     0.65      0.02     1263     0.02     0.02  Psych::ClassLoader#load
  2.11     0.67      0.02     1263     0.02     0.06  Psych::Visitors::ToRuby#deserialize
  2.11     0.69      0.02     1263     0.02     0.08  Psych::Visitors::ToRuby#visit_Psych_Nodes_Scalar
  2.11     0.71      0.02      695     0.03     0.13  Hash#inspect
  1.05     0.72      0.01      294     0.03     0.03  File#initialize
  1.05     0.73      0.01     1381     0.01     0.01  String#empty?
  1.05     0.74      0.01     1155     0.01     0.01  Hash#key?

Definitely usable :)

viroulep commented:
Updated and removed the `wip` label, so it's an actual PR :)

jonatanklosko commented on WcaOnRails/app/models/locale.rb:
```diff
@@ -0,0 +1,124 @@
+        # $1 is everything before the comment and the key
+        before = $1
+        # $2 is the hash(es) matched, or empty
+        hashes = $2
```
Actually, it's the comment lines matched.

jonatanklosko commented on WcaOnRails/app/models/locale.rb:
```diff
@@ -0,0 +1,124 @@
+    # Some keys are too long and would overflow from the list-item, shrink it to an arbitrary length
+    if key.length > 20
+      key = key[0...19] + "..."
+    end
```
Look at [this method](http://api.rubyonrails.org/classes/ActionView/Helpers/TextHelper.html#method-i-truncate) =)

jonatanklosko commented on WcaOnRails/app/models/locale.rb:
```diff
@@ -0,0 +1,124 @@
+        else
+          context.push(key)
+          missing_keys += get_all_recursive(value, context)
+          context.pop
```
You add the key temporarily, so what about doing it this way:
```ruby
missing_keys += get_all_recursive(value, [*context, key])
```

viroulep replied:
I didn't want to copy a whole array when I could pass a reference, but I'm fine with this way too :)

jonatanklosko commented on WcaOnRails/app/models/locale.rb:
```diff
@@ -0,0 +1,124 @@
+      else
+        context.push(key)
+        missing, outdated = compare_node_resursive(value, translation[key], context)
+        context.pop
```
Same as above =)

jonatanklosko commented on WcaOnRails/app/models/locale.rb:
```diff
@@ -0,0 +1,124 @@
+      else
+        context.push(key)
+        leaves += get_all_recursive(value, context)
+        context.pop
```
Same as above.

jonatanklosko commented on WcaOnRails/app/views/server_status/index.html.erb:
```diff
@@ -2,31 +2,131 @@
+                 else
+                   "success"
+                 end
+%>
```
This is kinda long, I'd suggest either doing it with ternary or moving this to a view helper.

jonatanklosko commented on WcaOnRails/app/views/server_status/index.html.erb:
```diff
@@ -2,31 +2,131 @@
+            ago and still has not run.
+            <%= @jobs_that_should_have_run_by_now.count %>
+            <%= "job".pluralize(@jobs_that_should_have_run_by_now.count) %>
+            <%= "is".pluralize(@jobs_that_should_have_run_by_now.count) %>
```
Could be a map & join, but whatever you prefer.

jonatanklosko commented on WcaOnRails/app/views/server_status/index.html.erb:
```diff
@@ -2,31 +2,131 @@
+         class_missing = "danger"
+         title = "Some items in the translation are missing and some other are outdated"
+       end
+    %>
```
Looks like a good candidate for a view helper =)

jonatanklosko commented on WcaOnRails/app/views/server_status/index.html.erb:
```diff
@@ -2,31 +2,131 @@
+          <div>
+        <% end %>
+            <span class="label label-<%= class_panel %>">i18n</span>
+            <span class="label label-<%= class_panel %>"><%= "#{l}" %></span>
```
I don't think you neee those `"#{}"` or `to_s` within `<%= %>`.

viroulep replied:
Right, I fixed it above too, as I use `l.to_s` in the same context.

jonatanklosko commented:
Thanks Philippe, LGTM! The page looks really clean =)
@jfly any thoughts?

jfly commented on WcaOnRails/app/helpers/server_status_helper.rb:
```diff
@@ -0,0 +1,22 @@
+      else
+        class_panel = class_outdated = "warning"
+        class_missing = "danger"
+        title = "Some items in the translation are missing and some other are outdated"
```
nit: "some other" -> "some others"

jfly replied:
Personal preference, but I'd probably have written this to make the 4 cases painfully clear:

```ruby
if missing_empty && outdated_empty
  ...
elsif missing_empty && !outdated_empty
  ...
elsif !missing_empty && outdated_empty
  ...
elsif !missing_empty && !outdated_empty
  ...
else
  fail
```

I acknowledge that a lot of people find this style pretty weird, but I like to show to the reader that I *did* explicitly consider every case, and I didn't forget about a possible state (hence the `else fail` that's impossible to hit) (Don't feel a need to respond to this.)

Lastly, I found the variable names `missing_empty` and `outdated_empty` a bit confusing here (does `missing_empty` mean "this is the empty that we're missing"?), but I cannot think of better names for them. Maybe `translations_are_missing` and `translations_are_outdated`?

jfly commented on WcaOnRails/app/models/locale.rb:
```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+class Locale < SimpleDelegator
+  include ActionView::Helpers::TextHelper
+  SOME_CHARS = '[\s\S]*?'
```
I must confess I don't understand this regex, doesn't `[\s\S]` cover every possible character?

jonatanklosko replied:
Yeah, it means really any possible character. IIRC it should work with a dot and the `m` flag, but I don't think it would be much better. Also, note that this one was made in JavaScript, where the above is the best solution.

viroulep replied:
Some numbers, constructing one translation locale (out of the rails structure), with the `Regexp::MULTILINE` enabled:
(I truncated the output when the calls added nothing to the cumulative time, therefore the last cumulative time is the total time)
<details>
  <summary>Using `.*`:</summary>

>  %   cumulative   self              self     total
 time   seconds   seconds    calls  ms/call  ms/call  name
 94.31     6.13      6.13      695     8.82     8.82  String#sub
  0.77     6.18      0.05      112     0.45     3.21  Kernel#require
  0.62     6.22      0.04     1392     0.03     0.85  Psych::Visitors::ToRuby#accept
  0.62     6.26      0.04     1263     0.03     0.07  Psych::Visitors::ToRuby#deserialize
  0.46     6.29      0.03     1267     0.02    38.43  Locale#decorate_with_hashes
  0.46     6.32      0.03     1392     0.02     0.81  Psych::Visitors::Visitor#visit
  0.46     6.35      0.03     1263     0.02     0.02  Psych::TreeBuilder#scalar
  0.31     6.37      0.02      657     0.03     0.05  Psych::ScalarScanner#tokenize
  0.31     6.39      0.02      695     0.03     0.03  Regexp#initialize
  0.15     6.40      0.01     1264     0.01     0.01  BasicObject#!
  0.15     6.41      0.01      130     0.08     0.08  Array#push
  0.15     6.42      0.01        1    10.00    60.00  Psych::Parser#parse
  0.15     6.43      0.01      667     0.01     0.01  Regexp#===
  0.15     6.44      0.01     1263     0.01     0.08  Psych::Visitors::ToRuby#visit_Psych_Nodes_Scalar
  0.15     6.45      0.01     1392     0.01     0.82  Psych::Visitors::Visitor#accept
  0.15     6.46      0.01      128     0.08     6.33  Psych::Visitors::ToRuby#visit_Psych_Nodes_Mapping
  0.15     6.47      0.01      128     0.08     0.08  Psych::TreeBuilder#end_mapping
  0.15     6.48      0.01     1263     0.01     0.02  Psych::ClassLoader#load
  0.15     6.49      0.01      695     0.01     0.01  Locale#leaf?
  0.15     6.50      0.01      128     0.08   190.23  Hash#each
</details>

<details>
  <summary>Using `.*?`:</summary>

>  %   cumulative   self              self     total
 time   seconds   seconds    calls  ms/call  ms/call  name
 22.81     0.13      0.13      695     0.19     0.20  String#sub
 12.28     0.20      0.07     1518     0.05     1.46  Locale#decorate_with_hashes
 10.53     0.26      0.06      823     0.07     1.88  Psych::Visitors::ToRuby#revive_hash
  7.02     0.30      0.04     1263     0.03     0.05  Psych::TreeBuilder#scalar
  7.02     0.34      0.04      112     0.36     2.68  Kernel#require
  5.26     0.37      0.03     1263     0.02     0.04  Psych::Visitors::ToRuby#visit_Psych_Nodes_Scalar
  5.26     0.40      0.03      695     0.04     0.04  Locale#leaf?
  3.51     0.42      0.02     2101     0.01     0.27  Class#new
  3.51     0.44      0.02     1706     0.01     0.01  Hash#[]=
  3.51     0.46      0.02     1392     0.01     0.74  Psych::Visitors::Visitor#visit
  1.75     0.47      0.01      196     0.05     3.98  Array#each
  1.75     0.48      0.01     1263     0.01     0.01  Psych::Nodes::Scalar#initialize
  1.75     0.49      0.01     1263     0.01     0.01  Psych::Visitors::ToRuby#resolve_class
  1.75     0.50      0.01     1263     0.01     0.02  Psych::Visitors::ToRuby#deserialize
  1.75     0.51      0.01     1392     0.01     0.75  Psych::Visitors::Visitor#accept
  1.75     0.52      0.01     1392     0.01     0.01  Hash#empty?
  1.75     0.53      0.01      695     0.01     0.01  Regexp#initialize
  1.75     0.54      0.01      128     0.08     0.08  Psych::TreeBuilder#start_mapping
  1.75     0.55      0.01     1391     0.01     0.01  Array#<<
  1.75     0.56      0.01        1    10.00    80.00  Psych::Parser#parse
  1.75     0.57      0.01     1392     0.01     0.76  Psych::Visitors::ToRuby#accept
</details>

<details>
<summary>Using `[.]*`:</summary>

>  %   cumulative   self              self     total
 time   seconds   seconds    calls  ms/call  ms/call  name
 62.90     0.78      0.78      695     1.12     1.14  String#sub
  4.84     0.84      0.06      825     0.07     8.93  Locale#decorate_with_hashes
  3.23     0.88      0.04     2101     0.02     0.59  Class#new
  2.42     0.91      0.03     1263     0.02     0.03  Psych::TreeBuilder#scalar
  2.42     0.94      0.03     1392     0.02     0.78  Psych::Visitors::Visitor#visit
  1.61     0.96      0.02      657     0.03     0.05  Psych::ScalarScanner#tokenize
  1.61     0.98      0.02      695     0.03     0.04  Locale#leaf?
  1.61     1.00      0.02      112     0.18     3.30  Kernel#require
  1.61     1.02      0.02      223     0.09     0.09  Gem.find_unresolved_default_spec
  1.61     1.04      0.02        1    20.00    70.00  Psych::Parser#parse
  1.61     1.06      0.02     2783     0.01     0.01  Hash#[]
  1.61     1.08      0.02     1263     0.02     0.06  Psych::Visitors::ToRuby#visit_Psych_Nodes_Scalar
  0.81     1.09      0.01       13     0.77     0.77  Array#map
  0.81     1.10      0.01     1264     0.01     0.01  BasicObject#!
  0.81     1.11      0.01     2087     0.00     0.00  Delegator.const_missing
  0.81     1.12      0.01      128     0.08     0.08  Psych::TreeBuilder#start_mapping
  0.81     1.13      0.01     1395     0.01     0.01  Kernel#hash
  0.81     1.14      0.01     1706     0.01     0.01  Hash#[]=
  0.81     1.15      0.01     1263     0.01     0.02  Psych::Visitors::ToRuby#resolve_class
  0.81     1.16      0.01      667     0.01     0.01  Regexp#===
  0.81     1.17      0.01     1392     0.01     0.80  Psych::Visitors::ToRuby#accept
  0.81     1.18      0.01      823     0.01     1.83  Psych::Visitors::ToRuby#revive_hash
  0.81     1.19      0.01      128     0.08     5.94  Enumerable#each_slice
  0.81     1.20      0.01      128     0.08     6.02  Psych::Visitors::ToRuby#visit_Psych_Nodes_Mapping
  0.81     1.21      0.01      695     0.01     0.01  Regexp#initialize
  0.81     1.22      0.01       34     0.29     0.29  nil#
  0.81     1.23      0.01     1392     0.01     0.79  Psych::Visitors::Visitor#accept
</details>

<details>
<summary>Using `[.]*?`:</summary>

>  %   cumulative   self              self     total
 time   seconds   seconds    calls  ms/call  ms/call  name
 70.63     0.89      0.89      695     1.28     1.28  String#sub
  3.97     0.94      0.05      825     0.06     9.53  Locale#decorate_with_hashes
  3.17     0.98      0.04      112     0.36     3.66  Kernel#require
  2.38     1.01      0.03      657     0.05     0.06  Psych::ScalarScanner#tokenize
  2.38     1.04      0.03     1263     0.02     0.02  Psych::TreeBuilder#scalar
  1.59     1.06      0.02      128     0.16     0.16  Psych::TreeBuilder#start_mapping
  1.59     1.08      0.02     1263     0.02     0.08  Psych::Visitors::ToRuby#deserialize
  1.59     1.10      0.02      823     0.02     1.80  Psych::Visitors::ToRuby#revive_hash
  1.59     1.12      0.02      196     0.10     3.88  Array#each
  1.59     1.14      0.02     1392     0.01     0.77  Psych::Visitors::Visitor#visit
  0.79     1.15      0.01      133     0.08     0.08  Array#last
  0.79     1.16      0.01        7     1.43     1.43  Gem::Specification.gemspec_stubs_in
  0.79     1.17      0.01     1099     0.01     0.01  Hash#key?
  0.79     1.18      0.01     1264     0.01     0.01  BasicObject#!
  0.79     1.19      0.01     1393     0.01     0.01  Kernel#class
  0.79     1.20      0.01        1    10.00    70.00  Psych::Parser#parse
  0.79     1.21      0.01     1263     0.01     0.02  Psych::ClassLoader#load
  0.79     1.22      0.01     1263     0.01     0.02  Psych::Visitors::ToRuby#resolve_class
  0.79     1.23      0.01     1263     0.01     0.09  Psych::Visitors::ToRuby#visit_Psych_Nodes_Scalar
  0.79     1.24      0.01      695     0.01     0.01  Locale#leaf?
  0.79     1.25      0.01      695     0.01     0.01  Regexp#initialize
  0.79     1.26      0.01     2783     0.00     0.00  Hash#[]
</details>

<details>
<summary>Our current solution (`[\s\S]*?` without multiline mode):</summary>

>  %   cumulative   self              self     total
 time   seconds   seconds    calls  ms/call  ms/call  name
 34.48     0.20      0.20      695     0.29     0.30  String#sub
 10.34     0.26      0.06     1518     0.04     1.55  Locale#decorate_with_hashes
  5.17     0.29      0.03     1263     0.02     0.03  Psych::TreeBuilder#scalar
  5.17     0.32      0.03     2783     0.01     0.01  Hash#[]
  5.17     0.35      0.03      657     0.05     0.06  Psych::ScalarScanner#tokenize
  5.17     0.38      0.03     1263     0.02     0.02  Psych::Visitors::ToRuby#resolve_class
  3.45     0.40      0.02     1392     0.01     0.75  Psych::Visitors::Visitor#visit
  3.45     0.42      0.02     1392     0.01     0.77  Psych::Visitors::ToRuby#accept
  3.45     0.44      0.02      823     0.02     1.79  Psych::Visitors::ToRuby#revive_hash
  1.72     0.45      0.01      196     0.05     3.78  Array#each
  1.72     0.46      0.01       19     0.53     0.53  Array#map
  1.72     0.47      0.01      112     0.09     3.39  Kernel#require
  1.72     0.48      0.01     1353     0.01     0.01  String#empty?
  1.72     0.49      0.01      128     0.08     0.16  Psych::TreeBuilder#start_mapping
  1.72     0.50      0.01     1263     0.01     0.07  Psych::Visitors::ToRuby#visit_Psych_Nodes_Scalar
  1.72     0.51      0.01      695     0.01     0.01  Locale#leaf?
  1.72     0.52      0.01       56     0.18     0.18  MonitorMixin#mon_enter
  1.72     0.53      0.01       30     0.33     0.33  Module#attr_accessor
  1.72     0.54      0.01     2101     0.00     0.27  Class#new
  1.72     0.55      0.01        6     1.67     1.67  Gem::Specification.dirs
  1.72     0.56      0.01      128     0.08     0.08  Psych::TreeBuilder#end_mapping
  1.72     0.57      0.01      130     0.08     0.08  Psych::Nodes::Node#initialize
  1.72     0.58      0.01      695     0.01     0.01  Regexp#initialize

</details>

I didn't realize making the match optional would make it so much faster!
I'll change the expression to `.*?`, it's the same execution time but should be less confusing.

jonatanklosko replied:
@viroulep What do you mean by the optional match? It's just non-greedy, and it's not for performance in this case, it won't work the other way.
Par example
```yaml
en:
  common:
    competitions: "Competitions"
  competitions:
    venue: "Venue"
```
If our regexps wouldn't use non-greedy matchers, then trying to match the key `en.common.competitions`, would match the `competitions:` tree.

viroulep replied:
Oh right, I actually tested with the actual French locale and saw no difference in the output, but everything is "outdated" so this "test" was actually a pretty bad one...
I also thought the `?` would make the `.*` optional, and that it could match 0 or 1 char multiple time (which is why I thought I could remove it). Very messed up interpretation actually...

viroulep replied:
As you will see, I changed it back to "[\s\S]*?" without the use of the regexp multiline mode.
I ran again into perf issues, I'm not sure if the regexp or the multiline caused them, but using the original expression works fast, so I must admit I not interested in investigating why and what caused the issues :p

jfly commented on WcaOnRails/app/views/server_status/index.html.erb:
```diff
@@ -2,31 +2,105 @@
     <%= alert :danger, "Check below, something may be wrong." %>
   <% end %>
+<% kind_class = @oldest_job_that_should_have_run_by_now ? "danger" : "success" %>
+    <div class="panel panel-<%= kind_class %>">
```
Is the indentation a bit screwy here?

jfly commented on WcaOnRails/app/views/server_status/index.html.erb:
```diff
@@ -2,31 +2,105 @@
+      <div class="panel-heading heading-as-link">
+        <h3 class="panel-title">
+        <% unless all_empty %>
+          <a data-toggle="collapse" href="#collapse-<%= l %>">
```
I'm not a big fan of ruby if statements interleaving with the heirarchy of the html like this. It feels unbalanced and hard to read.

Would it be possible to always make this a `<a data-toggle ...`, but to default to collapsed vs not collapsed based on `all_empty`?

viroulep replied:
My idea was to have no body to collapse if everything was empty, and using a div would make it explicit (by not changing the cursor) that it wasn't a link.

I found a better alternative using cursor in the css, I'll update this :)

viroulep replied:
@jfly Not sure if it was explicit enough, but I'll wait until you have a look at the "new" way of collapsing the panel before merging this ;)

jfly replied:
Just took a fresh look, I like it!

jfly commented on WcaOnRails/app/models/locale.rb:
```diff
@@ -0,0 +1,116 @@
+        next
+      end
+      if leaf?(value)
+        unless Digest::SHA1.hexdigest(value)[0..6] == translation[key]["_hash"]
```
I think this is worthy of a bit of an explanation. It would be really great if we documented our hash trick (maybe somewhere on our translations wiki?) for detecting out of date translations and then you could add a link to it here.

jonatanklosko commented:
@jfly *IMO* saving hashes as normal keys would be kinda dirty, I don't like this idea. The current one is nice despite its trickiness =)

viroulep commented:
Updated, I added another commit fixing some keys in our locale file: in Pedro's PR for the new locale I noticed `js-yaml` uses quotes on keys that could be parsed as integer.
It kind of messes up the locale comparison, and since we already do this for the events I changed the relevant keys in the locales to always quote integers.

jonatanklosko commented:
Still LGTM of course =)

jfly commented:
Still LGTM as well =)

viroulep commented:
Updated with support for pluralization, now such key is a leaf, and only it is reported as outdated, not the other ones:
![pluralized_keys](https://cloud.githubusercontent.com/assets/1007485/22076249/9ead4152-ddaf-11e6-84fe-adafc6f9aecd.png)

I also tested it with a freshly updated French locale, and it was able to detect correctly that it was up to date :)

jonatanklosko commented:
Lovely, LGTM *(still)* =)

viroulep commented:
Deployed, see [here](https://www.worldcubeassociation.org/server-status) for how it looks like when a locale is out of date!
